README.md
# Better JSONP [![npm](https://img.shields.io/npm/v/better-jsonp.svg)](https://www.npmjs.com/package/better-jsonp) [![Build Status](https://travis-ci.org/lbwa/jsonp.svg?branch=master)](https://travis-ci.org/lbwa/jsonp) [![npm](https://img.shields.io/npm/dt/better-jsonp.svg)](https://github.com/lbwa/jsonp)

A minimal and lightweight JSONP implementation which is used to be a kind of cross domain solutions.

- [See demo page]

- [Play on the playground]

[See demo page]:https://lbwa.github.io/jsonp/

[Play on the playground]:https://jsfiddle.net/608oagxd/8/

## Features

- Implement JSONP request from the browser

- Combine URL query parameters by default behavior

- Support the [Promise] API

- Limit JSONP request period

- Handle network error response

[Promise]:https://promisesaplus.com/

## Install

```bash
# using npm
npm i better-jsonp
```

```bash
# using yarn
yarn add better-jsonp
```

```html
<!-- using CDN -->
<!-- recommend you to add version number to url ending. eg.https://unpkg.com/better-jsonp@x.y.z -->
<script src="https://unpkg.com/better-jsonp@latest"></script>
<script src="https://cdn.jsdelivr.net/npm/better-jsonp@latest"></script>
```

## Promise polyfill

If your browser doesn't support [ES6 Promise], you should import polyfill to the global environment at the beginning of your application.

```js
require('es6-promise').polyfill()
```

***Notice***: Don't assign the result of `polyfill()` to any variable.

[es6 promise]:http://www.ecma-international.org/ecma-262/#sec-promise-objects

## Usage

JSONP ***only*** support GET methods, same as `better-JSONP`.

- The code below show you how to use package as a dependency

```js
// as a request dependency named jsonp
import jsonp from 'better-jsonp'
```

```js
jsonp({
  url: 'http://localhost',
  // global function named `callback` will be invoked when JSONP response
  jsonpCallback: 'callback', // any different name from request module
  timeout: 5000,
  // eg. ?customCallbackParams=...
  callbackParams: 'customCallbackParams',
  urlParams: {
    // eg. ?key0=0&key1=1...
    key0: 0,
    key1: 1
  }
})
  .then(res => console.log(res))
  .catch(err => console.error(err))
```

⚠️ ***Notice***: Parameter `jsonpCallback` value ***MUST NOT*** be same as request module name (eg. dependency named `jsonp` above code), otherwise request module only works once and function named value of parameter `jsonpCallback` will be reset to `null` (internal implementation) which means the same name request module will be also reset unexpectedly.

> For more customization capability, This package wouldn't convert `jsonpCallback` to a new name to prevent unexpected reset.

- You can also invoke function named `jsonp` directly in global environment if you have installed package from CDN.（[Play on the playground]）

```js
jsonp({
  // custom configuration
})
```

⚠️ ***Notice***: For same reason, parameter `jsonpCallback` value ***MUST NOT*** be `jsonp`.


## Parameters

| options parameter | type | required | description |
| ----------------- | ---- | -------- | ----------- |
|   `url`  | `String` |           true           | JSONP request address |
| `timeout` | `Number` | false, default : `6000` milliseconds | how long after timeout error is emitted. `0` to disable |
| `jsonpCallback`  | `String` | false, default : `'callback'+Date.now()` | global callback function name which is used to handle JSONP response. |
| `callbackParams` | `String` | false, default: `jsonpCallback` | name of query parameter to specify the callback name |
| `urlParams` |  `Object`  | false, default: `{}` | other parameters in query string parameters |

## Notice

- `Uncaught SyntaxError: Unexpected token :`error

It mostly doesn't support JSONP request when you are calling a JSON api. The difference between JSON api and JSONP is that  JSON api response with an object like `{ num: 1 }` (It will throw a error when client executed this response as a function. ). On the other hand, JSONP will respond with a function wrapped object like `jsonpCallback({ num: 1 })` and we will get what we need when client executed this response as a function.


CHANGELOG.md
<a name="1.1.3"></a>
## [1.1.3](https://github.com/lbwa/jsonp/compare/v1.1.2...v1.1.3) (2020-01-26)


### Bug Fixes

* manage global timer ([c7a277d](https://github.com/lbwa/jsonp/commit/c7a277d))
* Remove script from it's parentNode instead of _reference ([#2](https://github.com/lbwa/jsonp/issues/2)) ([fbdfcf5](https://github.com/lbwa/jsonp/commit/fbdfcf5))



<a name="1.1.2"></a>
## [1.1.2](https://github.com/lbwa/jsonp/compare/v1.1.1...v1.1.2) (2018-08-17)


### Bug Fixes

* **core:** remove unnecessary code for reduce code weight ([90bb90a](https://github.com/lbwa/jsonp/commit/90bb90a))



<a name="1.1.1"></a>
## [1.1.1](https://github.com/lbwa/jsonp/compare/v1.1.0...v1.1.1) (2018-08-17)


### Bug Fixes

* avoid global function named jsonp to reset unexpectedly ([98bf8a6](https://github.com/lbwa/jsonp/commit/98bf8a6))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/lbwa/jsonp/compare/v1.0.0...v1.1.0) (2018-06-20)


### Bug Fixes

* fix global variable ([b926279](https://github.com/lbwa/jsonp/commit/b926279))


### Features

* **core/Jsonp.js:** handle 404/500 ([346c54a](https://github.com/lbwa/jsonp/commit/346c54a))
* **lib/core/Jsonp.js:** draft of caught 404/500 ([cb957b2](https://github.com/lbwa/jsonp/commit/cb957b2))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/lbwa/jsonp/compare/090bbf3...v1.0.0) (2018-06-19)


### Features

* **lib/core/Jsonp.js lib/json.js samples/* publish.sh:** implement Promise API, deprecate `prefix` ([090bbf3](https://github.com/lbwa/jsonp/commit/090bbf3))


### BREAKING CHANGES

* **lib/core/Jsonp.js lib/json.js samples/* publish.sh:** implement Promise API, deprecate `prefix` option and `callback` option





lib/core/Jsonp.ts
import { options } from '../utils/types'
import { noop } from '../utils/index'

const PREFIX = 'callback'

export default class Jsonp {
  handler: Promise<object> // facade API based on error handler and response handler

  private _url: string // converted url based on basic url and other params
  private _jsonpCallback: string // response handler
  private _reference: Element // reference element
  private _script: HTMLScriptElement | null // trigger element
  private _timer: number

  constructor({
    url,
    timeout = 6000,
    jsonpCallback = `${PREFIX}${Date.now()}`,
    callbackParams = 'jsonpCallback',
    urlParams = {}
  }: options) {
    this.checkOptions({
      url,
      jsonpCallback
    })
    this.encodeURL({
      url,
      callbackParams,
      urlParams
    })
    this.initState({
      timeout,
      jsonpCallback
    })
    this.insert(this._url)
  }

  checkOptions({ url, jsonpCallback }: options) {
    if (!url) throw new Error('Please check your request url.')

    // Every jsonp request will reset global request function named value of
    // jsonpCallback, so this value MUST NOT be `jsonp`.

    // This checking only works in CDN installing, not as a dependency using
    if (jsonpCallback === 'jsonp')
      throw new Error(
        "Don't name jsonpCallback to `jsonp` for unexpected reset. Please use any non-jsonp value"
      )
  }

  initState({
    timeout,
    jsonpCallback
  }: {
    timeout: options['timeout']
    jsonpCallback: options['jsonpCallback']
  }) {
    this.createScript()

    // unique response handler
    this._jsonpCallback = jsonpCallback

    // keep behind setting this._jsonpCallback
    this.handler = this.createHandler()

    // set timer for limit request time
    this.createTimer(timeout)
  }

  createScript() {
    this._reference =
      document.getElementsByTagName('script')[0] ||
      document.body.lastElementChild
    this._script = document.createElement('script')
  }

  /**
   * 1. Request timer will be cleaned when response handler invoked.
   * 2. use arrow function to keep `this` keywords value (Jsonp instance).
   */
  createHandler() {
    return new Promise<object>((resolve, reject) => {
      // handle 404/500 in response
      this._script.onerror = () => {
        this.cleanScript()
        // clear timer
        if (this._timer) {
          window.clearTimeout(this._timer)
          this._timer = null
        }
        reject(
          new Error(
            `[SCRIPT ONERROR]: JSONP request unsuccessfully due to 404/500`
          )
        )
      }
      ;(window as any)[this._jsonpCallback] = (data: object) => {
        this.cleanScript()
        resolve(data)
      }
    })
  }

  // create a request timer for limiting request period
  createTimer(timeout: options['timeout']) {
    // It can be disable when param timeout equal falsy value (0, null etc.)
    if (timeout) {
      const id = window.setTimeout(() => {
        ;(window as any)[this._jsonpCallback] = noop
        this.cleanScript()
        this._timer = null
        throw new Error(
          '[TIMEOUT]: JSONP request unsuccessfully (eg.timeout or wrong url).'
        )
      }, timeout)
      this._timer = id
    }
  }

  encodeURL({ url, callbackParams, urlParams }: options) {
    // name of query parameter to specify the callback name
    // eg. ?callback=...
    const id = encodeURIComponent(this._jsonpCallback)
    url += `${url.indexOf('?') < 0 ? '?' : '&'}${callbackParams}=${id}`

    // add other parameters to url ending excluding callback name parameter
    const keys = Object.keys(urlParams)
    keys.forEach(key => {
      const value = urlParams[key] !== undefined ? urlParams[key] : ''
      url += `&${key}=${encodeURIComponent(value)}`
    })

    // converted request url
    this._url = url
  }

  // activate JSONP
  insert(url: string) {
    this._script.src = url
    this._reference.parentNode.insertBefore(this._script, this._reference)
  }

  cleanScript() {
    if (this._script && this._script.parentNode) {
      this._script.parentNode.removeChild(this._script)
      this._script = null
    }

    // reset response handler
    ;(window as any)[this._jsonpCallback] = noop
  }
}
